azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java

```diff
@@ -623,12 +629,28 @@ private boolean retryJobIfPossible(final ExecutableNode node) {
     }
   }
 
+  /**
+   * Recursively propagate status to parent flow. Alert on first error of the flow in new AZ
+   * dispatching design.
+   *
+   * @param base the base flow
+   * @param status the status to be propagated
+   */
   private void propagateStatus(final ExecutableFlowBase base, final Status status) {
     if (!Status.isStatusFinished(base.getStatus()) && base.getStatus() != Status.KILLING) {
       this.logger.info("Setting " + base.getNestedId() + " to " + status);
-      base.setStatus(status);
+      boolean shouldAlert = false;
```

Comment:
Have you considered separating the alerting part out into another method? method with one single responsibility brings more simplicity and composability.

Reply:
Thanks for the suggestion. I've also considered separating the alerting part. However, alerting happens when the flow status is propagated to `FAILED_FINISHING` for the first time. So it anyway needs to recursively call `getParentFlow` to access the root flow status, which is the same as what `propagateStatus` is doing. Also to make sure we only alert once, a global variable might be needed to keep the info whether alerting has happened or not. Considering the efficiency and simplicity of the code, I would prefer combining them together, maybe changing the method name to better describe the purpose of this method.

Reply:
Discussed offline and agreed on adding a new FLOW_STATUS_CHANGED event type and switching to the event-driven model in the future.